src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,290 @@
+};
+use nix;
+use safe_path;
+use std::{fs, path::PathBuf, time::Duration};
+use url;
+
+use runtimes;
+
+const TIMEOUT: Duration = Duration::new(5, 0);
+const CONTENT_TYPE_JSON: &str = "application/json";
+
+pub fn handle_direct_volume(vol_cmd: DirectVolumeCommand) -> Result<()> {
+    if !nix::unistd::Uid::effective().is_root() {
+        return Err(anyhow!("The Direct-Volume Subcommands are privileged, please retry with privileged permissions"));
+    }
+    let command = vol_cmd.directvol_cmd;
+    let cmd_result: Option<String> = match command {
+        DirectVolSubcommand::Add(args) => add(&args.volume_path, &args.mount_info)?,
+        DirectVolSubcommand::Remove(args) => remove(&args.volume_path)?,
+        DirectVolSubcommand::Stats(args) => executor::block_on(stats(&args.volume_path))?,
+        DirectVolSubcommand::Resize(args) => {
+            executor::block_on(resize(&args.volume_path, args.resize_size))?
+        }
+    };
+    if let Some(cmd_result) = cmd_result {
+        println!("{:?}", cmd_result);
+    }
+    Ok(())
+}
+
+async fn resize(volume_path: &str, size: u64) -> Result<Option<String>> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volume_mount_info(volume_path)?;
+    let resize_req = runtimes::ResizeRequest {
+        size,
+        volume_path: mount_info.device,
+    };
+    let encoded = serde_json::to_string(&resize_req)?;
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT))?;
+
+    let url = runtimes::DIRECT_VOLUMN_RESIZE_URL;
+    let response = shim_client
+        .post(url, &String::from(CONTENT_TYPE_JSON), &encoded)
+        .await?;
+    let status = response.status();
+    if status != hyper::StatusCode::OK {
+        let body = hyper::body::to_bytes(response.into_body()).await?;
+        let body = String::from_utf8(body.to_vec())?;
+        return Err(anyhow!(
+            "failed to resize volume ({:?}): {:?}",
+            status,
+            body
+        ));
+    }
+    Ok(None)
+}
+
+async fn stats(volume_path: &str) -> Result<Option<String>> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volume_mount_info(volume_path)?;
+
+    let req_url =
+        url::form_urlencoded::Serializer::new(String::from(runtimes::DIRECT_VOLUMN_STATS_URL))
+            .append_pair(runtimes::DIRECT_VOLUMN_PATH_KEY, &mount_info.device)
+            .finish();
+
+    // sid: sandbox id
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT))?;
+    let response = shim_client.get(&req_url).await?;
+    // turn body into string
+    let body = hyper::body::to_bytes(response.into_body()).await?;
+    let body = String::from_utf8(body.to_vec())?;
+    Ok(Some(body))
+}
+
+// join_path joins user provided volumepath with kata direct-volume root path
+fn join_path(prefix: &str, volume_path: &str) -> Result<PathBuf> {
```

Comment:
Could you add some table-based unit tests for this function?

Reply:
I think this is added in https://github.com/kata-containers/kata-containers/pull/5467/commits/9b01957f8fc23bf60e5c2292763b3bd93185a90e 😀